hazelcast-client-new/src/test/java/com/hazelcast/client/listeners/ListenerLeakTest.java

```diff
@@ -0,0 +1,224 @@
+import com.hazelcast.core.EntryAdapter;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IFunction;
+import com.hazelcast.core.IList;
+import com.hazelcast.core.IMap;
+import com.hazelcast.core.IQueue;
+import com.hazelcast.core.ISet;
+import com.hazelcast.core.ITopic;
+import com.hazelcast.core.Message;
+import com.hazelcast.core.MessageListener;
+import com.hazelcast.core.MultiMap;
+import com.hazelcast.core.ReplicatedMap;
+import com.hazelcast.instance.Node;
+import com.hazelcast.map.MapPartitionLostEvent;
+import com.hazelcast.map.listener.MapListener;
+import com.hazelcast.map.listener.MapPartitionLostListener;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.Collection;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class ListenerLeakTest extends HazelcastTestSupport {
+
+    private final TestHazelcastFactory hazelcastFactory = new TestHazelcastFactory();
+
+    @After
+    public void cleanup() {
+        hazelcastFactory.terminateAll();
+    }
+
+    private void testListenerLeak(IFunction listenerAddRemoveFunc) {
+        HazelcastInstance hazelcast = hazelcastFactory.newHazelcastInstance();
+        Node node = getNode(hazelcast);
+
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient();
+        String id = (String) listenerAddRemoveFunc.apply(client);
+
+        Collection<ClientEndpoint> endpoints = node.clientEngine.getEndpointManager().getEndpoints();
+        for (ClientEndpoint endpoint : endpoints) {
+            assertFalse(endpoint.removeListenerRegistration(id));
+        }
+    }
+
+    @Test
+    public void testMapEntryListeners() {
+        testListenerLeak(new IFunction() {
+            @Override
+            public Object apply(Object input) {
+                HazelcastInstance client = (HazelcastInstance) input;
+                IMap map = client.getMap(randomString());
+                String id = map.addEntryListener(new MapListener() {
+                }, false);
+                assertTrue(map.removeEntryListener(id));
+                return id;
+            }
+        });
+    }
+
+    @Test
+    public void testMapPartitionLostListeners() {
+        testListenerLeak(new IFunction() {
+            @Override
+            public Object apply(Object input) {
+                HazelcastInstance client = (HazelcastInstance) input;
+                IMap map = client.getMap(randomString());
+                String id = map.addPartitionLostListener(new MapPartitionLostListener() {
```

Comment:
Use a mock

 String id = map.addPartitionLostListener(mock(MapPartitionLostListener.class));

Easier to read